lib/3scale/client.rb

```diff
@@ -122,20 +124,27 @@ def authrep(options)
     #
     # == Examples
     #
-    #   # Report two transactions of two applications.
-    #   client.report({:app_id => 'foo', :usage => {'hits' => 1}},
-    #                 {:app_id => 'bar', :usage => {'hits' => 1}})
+    #    Report two transactions of two applications. Using the default service.
+    #    client.report(transactions: [{:app_id => 'foo', :usage => {'hits' => 1}},
+    #                                 {:app_id => 'bar', :usage => {'hits' => 1}}])
     #
-    #   # Report one transaction with timestamp.
-    #   client.report({:app_id    => 'foo',
-    #                  :timestamp => Time.local(2010, 4, 27, 15, 14),
-    #                  :usage     => {'hits' => 1})
+    #    Report one transaction with timestamp. Using the default service.
+    #    client.report(transactions: [{:app_id    => 'foo',
+    #                                  :timestamp => Time.local(2010, 4, 27, 15, 14),
+    #                                  :usage     => {'hits' => 1}])
     #
-    def report(*transactions)
-      raise ArgumentError, 'no transactions to report' if transactions.empty?
+    #    Report a transaction specifying the service.
+    #    client.report(transactions: [{:app_id => 'foo', :usage => {'hits' => 1}}],
+    #                  service_id: 'a_service_id')
+    #
+    def report(transactions: [], service_id: nil)
```

Comment:
Some comments: 

1) Agree with @mikz that the best way would be to keep the solution backward compatible, but add a deprecation warning and encourage everybody use the `(transactions: [], service_id: 123)` interface.

2) The solution:

    def report(*reports, transactions: [], service_id: nil)
        ...
    end

doesn't work for the previous interface, for example: `report({app_id: 1, usage: 2}, {app_id: 3, usage: 4})`.
It fails with `ArgumentError: unknown keywords: app_id, usage`, other than that a really nice solution

3) Regarding the solution:
    def report(*transactions)
        *reports, service_id = transactions
        unless service_id.is_a?(Numeric)
        ...
    end

To be honest I don't like the interface too much (an optional numeric value as the last element of the list).

4) An alternative solution:

    def report(*params)
      if params.first.has_key?(:transactions)
        transactions = params.first[:transactions]
        service_id = params.first[:service_id]
      else
        transactions = params
        warn 'deprecation warning'
      end
      # here the `transactions` has the array of reports
      ...
    end

Not very elegant... But it will work with all the following:


    report({app_id: 1, usage: 2}, {app_id: 3, usage: 4})
    report({app_id: 1, usage:2})
    report({transactions: [{app_id: 1, usage: 2}, {app_id: 3, usage: 4}], service_id: 111})
    report(transactions: [{app_id: 1, usage: 2}, {app_id: 3, usage: 4}])

5) I don't know if there is a problem to drop <= 2.0 compatibility... We don't know if any of the existing customers user it.